pkg/services/sqlstore/migrations/ualert/tokenizer.go

```diff
@@ -0,0 +1,139 @@
+// This file contains code that parses templates from old alerting into a sequence
+// of tokens. Each token can be either a string literal or a variable.
+
+package ualert
+
+import (
+	"errors"
+	"fmt"
+	"unicode"
+)
+
+var (
+	ErrEOF = errors.New("EOF")
+)
+
+// Token contains either a whitespace character, such as a space, tab, or newline;
+// a string literal; or a variable.
+type Token struct {
+	Literal  string
+	Variable string
+}
+
+func (t Token) IsLiteral() bool {
+	return t.Literal != ""
+}
+
+func (t Token) IsVariable() bool {
+	return t.Variable != ""
+}
+
+func (t Token) String() string {
+	if t.IsLiteral() {
+		return t.Literal
+	} else if t.IsVariable() {
+		return t.Variable
+	} else {
+		panic("empty token")
+	}
+}
+
+func tokenizeLiteral(in []rune) (Token, int, error) {
+	var (
+		pos   int
+		r     rune
+		runes []rune
+	)
+
+	// consume runes until the first '$' or the end of in
+	for pos < len(in) {
+		if r = in[pos]; r == '$' {
+			// don't consume the '$' as this is the start of a variable
+			break
+		}
+		runes = append(runes, r)
```

Comment:
we can avoid this by just taking sub-slice. We know start position and end position.

Reply:
That really simplifies this yeah, good catch.